File: aes/aes-256-cbc.ts
import {
    BinaryLike,
    createCipheriv,
    createDecipheriv,
    createHmac,
    pbkdf2Sync,
    randomBytes,
    timingSafeEqual
} from 'crypto'

export class AES256CBC {
    public encrypt(text: string, masterkey: BinaryLike) {
        const iv = randomBytes(16)
        const salt = randomBytes(64)

        const key = pbkdf2Sync(masterkey, salt, 2145, 32, 'sha512')
        const cipher = createCipheriv('aes-256-cbc', key, iv)

        let encrypted = Buffer.concat([cipher.update(text, 'utf-8'), cipher.final()])

        const hmac = createHmac('sha256', key)
        hmac.update(encrypted)

        return Buffer.concat([salt, iv, encrypted, hmac.digest()]).toString('base64')
    }

    public decrypt(encdata: string, masterkey: BinaryLike) {
        const bData = Buffer.from(encdata, 'base64')

        const salt = bData.slice(0, 64)
        const iv = bData.slice(64, 80)
        const encrypted = bData.slice(80, -32)
        const receivedHmac = bData.slice(-32)

        const key = pbkdf2Sync(masterkey, salt, 2145, 32, 'sha512')

        const hmac = createHmac('sha256', key)
        hmac.update(encrypted)

        const calculatedHmac = hmac.digest()

        if (!timingSafeEqual(calculatedHmac, receivedHmac)) {
            throw new Error('HMAC verification failed')
        }

        const decipher = createDecipheriv('aes-256-cbc', key, iv)

        const decrypted = decipher.update(encrypted) + decipher.final('utf-8')

        return decrypted
    }
}

File: py/aes_256_cbc.py
import base64
import hashlib
from Crypto.Cipher import AES
from Crypto.Random import get_random_bytes
from Crypto.Util.Padding import pad, unpad
from Crypto.Hash import HMAC


class AES256CBC:
    def encrypt(self, text: str, masterkey: str) -> str:
        iv = get_random_bytes(16)
        salt = get_random_bytes(64)

        key = hashlib.pbkdf2_hmac(
            "sha512", masterkey.encode("utf-8"), salt, 2145, dklen=32
        )

        cipher = AES.new(key, AES.MODE_CBC, iv)

        padded_text = pad(text.encode("utf-8"), AES.block_size)
        encrypted = cipher.encrypt(padded_text)

        hmac = HMAC.new(key, encrypted, hashlib.sha256)
        hmac_digest = hmac.digest()

        encrypted_data = salt + iv + encrypted + hmac_digest

        return base64.b64encode(encrypted_data).decode("utf-8")

    def decrypt(self, encdata: str, masterkey: str) -> str:
        bData = base64.b64decode(encdata)

        salt = bData[:64]
        iv = bData[64:80]
        encrypted = bData[80:-32]
        received_hmac = bData[-32:]

        key = hashlib.pbkdf2_hmac(
            "sha512", masterkey.encode("utf-8"), salt, 2145, dklen=32
        )

        hmac = HMAC.new(key, encrypted, hashlib.sha256)
        calculated_hmac = hmac.digest()

        if calculated_hmac != received_hmac:
            raise ValueError("HMAC verification failed")

        cipher = AES.new(key, AES.MODE_CBC, iv)
        decrypted = cipher.decrypt(encrypted)
        unpadded_text = unpad(decrypted, AES.block_size)

        return unpadded_text.decode("utf-8")
